### ISSUE TITLE
Service not found error preventing spawn

### ISSUE THREAD

**Issue by hdlbq** [jupyterhub/dockerspawner#343]
I installed a jupterhub about one year ago, and it was working well. The hub runs in a docker, as the nginx proxy that allows an https access. The other nodes are managed in a docker swarm. The users are authenticated using CAS. The home directories of the users are stored on a nfs-mounted volume.

Randomly, users can't login, and the log shows an error 403:
```
[W 2019-10-18 11:30:53.464 JupyterHub log:174] 403 POST /hub/api/users/some-user/activity
```

I should say that the hub has been booted hours ago, starting from an empty jupyterhub.sqlite file, never restarted, and the sqlite file left untouched. The error seems to appear more frequently when a user stops and restarts his server. The configuration is the same for all users, but the error is frequent for some users, rare for other users.
Moreover, the error seems to disappear when I run the hub without any worker nodes in the swarm (all spawned servers running on the hub's host). My own account (admin) seems not to be affected by the error.

Note: I submited this issue to the jupyterhub/jupyterhub gitlab project. 
minrk answered:
> 
> The 403 errors here are caused by the API request made by single-user servers to register activity. Where there's definitely a bug (in SwarmSpawner, it looks like), is in the fact we see logs of the Spawner claiming that the service was not started, but it clearly was because it is making API requests to the Hub.
> 
> The sequence of events:
> 
>     SwarmSpawner launches the service
>     (bug somewhere) SwarmSpawner believes this has failed, but it has not.
>     Hub begins cleanup of server, including revoking credentials for the API token allocated to the server.
>     the service finishes starting, and starts making API requests, but its token has been revoked in step 3, resulting in 403.
> 
> So the 403 is a symptom, but the real error is the SwarmSpawner is starting servers, but it thinks it is failing somehow.

```
| [I 2019-10-22 12:40:53.176 JupyterHub base:812] User userc-testc took 53.770 seconds to start
```

the error seems to be related to the following sequence of events:
```
 [D 2019-10-18 11:29:59.979 JupyterHub user:542] Calling Spawner.start for kaniav-kamary
 [D 2019-10-18 11:29:59.988 JupyterHub dockerspawner:813] Getting container 'jupyter-kaniav-kamary' for dockerspawner::start before remove id:
 [I 2019-10-18 11:29:59.994 JupyterHub dockerspawner:820] Service 'jupyter-kaniav-kamary' is gone
 [I 2019-10-18 11:30:00.016 JupyterHub dockerspawner:1030] Created service jupyter-kaniav-kamary (id: ge4p07y) from image 160.228.22.168:5000/hdlbq/cs-notebook-r
 [I 2019-10-18 11:30:00.017 JupyterHub dockerspawner:1053] Starting service jupyter-kaniav-kamary (id: ge4p07y)
 [D 2019-10-18 11:30:00.017 JupyterHub swarmspawner:144] Getting task of service 'jupyter-kaniav-kamary'
| [D 2019-10-18 11:30:00.017 JupyterHub dockerspawner:813] Getting container 'jupyter-kaniav-kamary' for swarmspawner::get_task id:ge4p07y
| [D 2019-10-18 11:30:00.026 JupyterHub swarmspawner:256] Service ge4p07y state: pending
| [I 2019-10-18 11:30:00.920 JupyterHub log:174] 302 GET /hub/spawn -> /hub/spawn-pending/kaniav-kamary (kaniav-kamary@10.255.0.2) 1019.67ms
| [I 2019-10-18 11:30:00.960 JupyterHub pages:303] kaniav-kamary is pending spawn
| [I 2019-10-18 11:30:00.962 JupyterHub log:174] 200 GET /hub/spawn-pending/kaniav-kamary (kaniav-kamary@10.255.0.2) 18.15ms
| [D 2019-10-18 11:30:01.027 JupyterHub swarmspawner:144] Getting task of service 'jupyter-kaniav-kamary'
| [D 2019-10-18 11:30:01.028 JupyterHub dockerspawner:813] Getting container 'jupyter-kaniav-kamary' for swarmspawner::get_task id:ge4p07y
| [E 2019-10-18 11:30:01.036 JupyterHub user:626] Unhandled error starting kaniav-kamary's server: Service jupyter-kaniav-kamary not found
| [D 2019-10-18 11:30:01.036 JupyterHub user:724] Stopping kaniav-kamary
| [D 2019-10-18 11:30:01.036 JupyterHub swarmspawner:144] Getting task of service 'jupyter-kaniav-kamary'
| [D 2019-10-18 11:30:01.037 JupyterHub dockerspawner:813] Getting container 'jupyter-kaniav-kamary' for swarmspawner::get_task id:ge4p07y
| [W 2019-10-18 11:30:01.044 JupyterHub swarmspawner:128] Service jupyter-kaniav-kamary not found
| [D 2019-10-18 11:30:01.057 JupyterHub user:752] Deleting oauth client jupyterhub-user-kaniav-kamary
| [D 2019-10-18 11:30:01.069 JupyterHub user:755] Finished stopping kaniav-kamary
| ERROR:asyncio:Task exception was never retrieved
| future: <Task finished coro=<BaseHandler.spawn_single_user() done, defined at /opt/conda/lib/python3.6/site-packages/jupyterhub/handlers/base.py:697> exception=RuntimeError('Service jupyter-kaniav-kamary not found',)>
| Traceback (most recent call last):
|   File "/opt/conda/lib/python3.6/site-packages/jupyterhub/handlers/base.py", line 889, in spawn_single_user
|     timedelta(seconds=self.slow_spawn_timeout), finish_spawn_future
|   File "/opt/conda/lib/python3.6/site-packages/jupyterhub/handlers/base.py", line 807, in finish_user_spawn
|     await spawn_future
|   File "/opt/conda/lib/python3.6/site-packages/jupyterhub/user.py", line 642, in spawn
|     raise e
|   File "/opt/conda/lib/python3.6/site-packages/jupyterhub/user.py", line 546, in spawn
|     url = await gen.with_timeout(timedelta(seconds=spawner.start_timeout), f)
|   File "/opt/conda/lib/python3.6/site-packages/tornado/gen.py", line 736, in run
|     yielded = self.gen.throw(*exc_info)  # type: ignore
|   File "/opt/conda/lib/python3.6/site-packages/dockerspawner/dockerspawner.py", line 1057, in start
|     yield self.start_object()
|   File "/opt/conda/lib/python3.6/site-packages/tornado/gen.py", line 729, in run
|   File "/opt/conda/lib/python3.6/site-packages/tornado/gen.py", line 742, in run
|     yielded = self.gen.send(value)
|   File "/opt/conda/lib/python3.6/site-packages/dockerspawner/swarmspawner.py", line 252, in start_object
|     raise RuntimeError("Service %s not found" % self.service_name)
| RuntimeError: Service jupyter-kaniav-kamary not found
| [I 2019-10-18 11:30:01.107 JupyterHub log:174] 200 GET /hub/api/users/kaniav-kamary/server/progress (kaniav-kamary@10.255.0.2) 16.13ms
| [W 2019-10-18 11:30:53.464 JupyterHub log:174] 403 POST /hub/api/users/kaniav-kamary/activity (@10.0.0.48) 3.26ms
| [W 2019-10-18 11:30:54.423 JupyterHub log:174] 403 POST /hub/api/users/kaniav-kamary/activity (@10.0.0.48) 3.25ms
| [W 2019-10-18 11:30:55.849 JupyterHub log:174] 403 POST /hub/api/users/kaniav-kamary/activity (@10.0.0.48) 3.77ms
| [W 2019-10-18 11:30:59.051 JupyterHub log:174] 403 POST /hub/api/users/kaniav-kamary/activity (@10.0.0.48) 3.36ms
```

```
Debian 9.11
Docker version 19.03.3, build a872fc2f86
jupyterhub docker version: latest
nginx docker version: latest
```

**Comment by hdlbq** [comment]
Hello,
this issue seems to be linked to the "Timeout potentially due to docker inspect_container prior to creation #330"

**Comment by hdlbq** [comment]
Unfortunately, the problem is more rare, but still present:
```
 [D 2019-10-29 17:14:41.094 JupyterHub user:542] Calling Spawner.start for firstname-lastname
 [D 2019-10-29 17:14:41.103 JupyterHub dockerspawner:813] Getting container 'jupyter-firstname-lastname' for dockerspawner::start before remove id:cewue4q
 [W 2019-10-29 17:14:41.108 JupyterHub dockerspawner:1016] Removing service that should have been cleaned up: jupyter-firstname-lastname (id: cewue4q)
 [I 2019-10-29 17:14:41.108 JupyterHub swarmspawner:239] Removing service cewue4q66bfyfund4lrp8va4a
 [I 2019-10-29 17:14:41.171 JupyterHub dockerspawner:1030] Created service jupyter-firstname-lastname (id: nrjkrbw) from image 160.228.22.168:5000/hdlbq/cs-notebook-r:latest
 [I 2019-10-29 17:14:41.171 JupyterHub dockerspawner:1053] Starting service jupyter-firstname-lastname (id: nrjkrbw)
 [D 2019-10-29 17:14:41.171 JupyterHub swarmspawner:144] Getting task of service 'jupyter-firstname-lastname'
 [D 2019-10-29 17:14:41.172 JupyterHub dockerspawner:813] Getting container 'jupyter-firstname-lastname' for swarmspawner::get_task id:nrjkrbw
 [E 2019-10-29 17:14:41.183 JupyterHub user:626] Unhandled error starting firstname-lastname's server: Service jupyter-firstname-lastname not found
 [D 2019-10-29 17:14:41.183 JupyterHub user:724] Stopping firstname-lastname
 [D 2019-10-29 17:14:41.184 JupyterHub swarmspawner:144] Getting task of service 'jupyter-firstname-lastname'
 [D 2019-10-29 17:14:41.184 JupyterHub dockerspawner:813] Getting container 'jupyter-firstname-lastname' for swarmspawner::get_task id:nrjkrbw
 [W 2019-10-29 17:14:41.194 JupyterHub swarmspawner:128] Service jupyter-firstname-lastname not found
 [D 2019-10-29 17:14:41.207 JupyterHub user:752] Deleting oauth client jupyterhub-user-firstname-lastname
 [D 2019-10-29 17:14:41.240 JupyterHub user:755] Finished stopping firstname-lastname
 [I 2019-10-29 17:14:41.273 JupyterHub log:174] 302 GET /hub/spawn -> /hub/spawn-pending/firstname-lastname (firstname-lastname@10.255.0.2) 256.50ms
 ERROR:asyncio:Task exception was never retrieved
 future: <Task finished coro=<BaseHandler.spawn_single_user() done, defined at /opt/conda/lib/python3.6/site-packages/jupyterhub/handlers/base.py:697> exception=RuntimeE
rror('Service jupyter-firstname-lastname not found',)>
 Traceback (most recent call last):
   File "/opt/conda/lib/python3.6/site-packages/jupyterhub/handlers/base.py", line 889, in spawn_single_user
     timedelta(seconds=self.slow_spawn_timeout), finish_spawn_future
   File "/opt/conda/lib/python3.6/site-packages/jupyterhub/handlers/base.py", line 807, in finish_user_spawn
     await spawn_future
   File "/opt/conda/lib/python3.6/site-packages/jupyterhub/user.py", line 642, in spawn
     raise e
   File "/opt/conda/lib/python3.6/site-packages/jupyterhub/user.py", line 546, in spawn
     url = await gen.with_timeout(timedelta(seconds=spawner.start_timeout), f)
   File "/opt/conda/lib/python3.6/site-packages/tornado/gen.py", line 736, in run
     yielded = self.gen.throw(*exc_info)  # type: ignore
   File "/opt/conda/lib/python3.6/site-packages/dockerspawner/dockerspawner.py", line 1057, in start
     yield self.start_object()
   File "/opt/conda/lib/python3.6/site-packages/tornado/gen.py", line 729, in run
     value = future.result()
   File "/opt/conda/lib/python3.6/site-packages/tornado/gen.py", line 742, in run
     yielded = self.gen.send(value)
   File "/opt/conda/lib/python3.6/site-packages/dockerspawner/swarmspawner.py", line 259, in start_object
     raise RuntimeError("Service %s not found" % self.service_name)
 RuntimeError: Service jupyter-firstname-lastname not found
 [E 2019-10-29 17:14:41.317 JupyterHub pages:284] Previous spawn for firstname-lastname failed: Service jupyter-firstname-lastname not found
 [E 2019-10-29 17:14:41.319 JupyterHub log:166] {
       "Cookie": "jupyterhub-hub-login=[secret]; jupyterhub-session-id=[secret]; _xsrf=[secret]; _gid=[secret]; _ga=[secret]",
       "Accept-Encoding": "gzip, deflate, br",
       "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.102 Safari/537.36 Edge/18.18362",
       "Upgrade-Insecure-Requests": "1",
       "Accept-Language": "fr-FR",
       "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
       "Cache-Control": "max-age=0",
       "Referer": "",
       "Connection": "close",
       "X-Nginx-Proxy": "true",
       "X-Forwarded-For": "10.255.0.2",
       "Host": "jupyter.centralesupelec.fr",
       "X-Real-Ip": "10.255.0.2"
     }
 [E 2019-10-29 17:14:41.319 JupyterHub log:174] 500 GET /hub/spawn-pending/firstname-lastname (firstname-lastname@10.255.0.2) 21.42ms
 [D 2019-10-29 17:14:41.372 JupyterHub swarmspawner:133] Service dijqif3 status: {'ContainerStatus': {'ContainerID': '178c0b2bff6e802726c31133892138c8ed4e51a19b19eb6e2dbb96b23f1e7a52',
                          'ExitCode': -1,
                          'PID': 21195},
      'Message': 'started',
      'PortStatus': {},
      'State': 'running',
      'Timestamp': '2019-10-29T15:32:07.734537946Z'}
 [D 2019-10-29 17:14:41.528 JupyterHub log:174] 200 GET /hub/static/css/style.min.css?v=883a1cf654bd9f5cdc5ca7c91efe7d71 (@10.255.0.2) 2.16ms
 [D 2019-10-29 17:14:41.578 JupyterHub log:174] 200 GET /hub/static/components/bootstrap/dist/js/bootstrap.min.js?v=2f34b630ffe30ba2ff2b91e3f3c322a1 (@10.255.0.2) 4.39ms
 [D 2019-10-29 17:14:41.582 JupyterHub log:174] 200 GET /hub/static/components/requirejs/require.js?v=f0cc8bbb2fcef87fc194fecbb632fcfa (@10.255.0.2) 7.09ms
 [D 2019-10-29 17:14:41.583 JupyterHub log:174] 200 GET /hub/logo (@10.255.0.2) 7.68ms
 [D 2019-10-29 17:14:41.585 JupyterHub log:174] 200 GET /hub/static/components/jquery/dist/jquery.min.js?v=220afd743d9e9643852e31a135a9f3ae (@10.255.0.2) 8.83ms
 [D 2019-10-29 17:14:42.034 JupyterHub log:174] 200 GET /hub/static/components/font-awesome/fonts/fontawesome-webfont.woff2?v=4.7.0 (@10.255.0.2) 2.68ms
 [I 2019-10-29 17:14:43.392 JupyterHub log:174] 200 GET /hub/api (@10.0.0.3) 1.40ms
 [W 2019-10-29 17:14:43.406 JupyterHub log:174] 403 POST /hub/api/users/firstname-lastname/activity (@10.0.0.3) 4.44ms
 [W 2019-10-29 17:14:44.009 JupyterHub log:174] 403 POST /hub/api/users/firstname-lastname/activity (@10.0.0.3) 3.55ms
 [I 2019-10-29 17:14:44.096 JupyterHub log:174] 200 GET /hub/api (@10.0.0.6) 0.92ms
 [W 2019-10-29 17:14:44.108 JupyterHub log:174] 403 POST /hub/api/users/firstname-lastname/activity (@10.0.0.6) 3.10ms
 [D 2019-10-29 17:14:44.217 JupyterHub pages:165] Triggering spawn with default options for firstname-lastname
 [D 2019-10-29 17:14:44.217 JupyterHub base:780] Initiating spawn for firstname-lastname
```

**Comment by zhiyuli** [comment]
I am having the same issue. The above patch does reduce its occurrence but still happens in some cases.

**Comment by hdlbq** [comment]
Hi,
Yes, you're right. It seems that the error occurs:
- more frequently if you have two swarm managers
- very rarely if you don't have additional node in the swarm (just the hub)
I think that the error is link to the responsiveness of the swarm.

**Comment by RonnieDilli** [comment]
> Hi,
> Yes, you're right. It seems that the error occurs:
> 
> * more frequently if you have two swarm managers
> * very rarely if you don't have additional node in the swarm (just the hub)
>   I think that the error is linked to the responsiveness of the swarm.
>   I will investigate deeper, but it is rather hard since I have only one "big" cluster, and a two-vm test cluster, with only two users: me and a test account. Moreover, retro-engineering the code in a context of asynchronous calls (that leads to an arbitrary order of log lines) is not easy.

We are facing the same problem with our Swarm, and your points seem to have a real impact.
In production deployment we have a setup like the one you listed: two swarm managers and other applications running along side in them.
With a simpler setup in our test environment, with only one manager, the problem is almost never seen.

Any permanent solutions for this?

**Comment by hdlbq** [comment]
Hi,
I'm working on two axis:
- switching to  JupyterHub 1.2.0dev 20200918085642, but I'm facing to another issue (https://github.com/jupyterhub/jupyterhub/issues/3182). This version seems to reduce our problem.
- redefining the 500 and 503 error templates, to redirect the user to his home. It's clearly not an answer, but it tends to minimize the impact on him while waiting for a patch.

**Comment by minrk** [comment]
I think this was closed by #396. Please feel free to open a new Issue if you are still facing this problem.
